src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -30,17 +30,17 @@ public partial class Solution
         // Values for all these are created on demand.
         private ImmutableHashMap<ProjectId, Project> _projectIdToProjectMap;
 
+        private readonly SemaphoreSlim _gate = new(initialCount: 1);
```

Comment:
I think I'd appreciate a comment here either what this is guarding, or leave a note on _cachedFrozenSolution that it's guarded by this.